File: tests/wipcAuthSession.test.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { describe, expect, it } from "vitest";
import {
  WipcProtocolError,
  deriveWipcParentProof
} from "../src/shared/wipc.js";
import {
  WipcHelperAuthSession,
  WipcParentAuthSession
} from "../src/shared/wipcAuthSession.js";
import { createWipcAuthSessionV1Fixture } from "../src/shared/wipcContract.js";
import { serializeCanonicalContractJson } from "../src/shared/schemas/remoteProtocolContract.js";

const repositoryRoot = path.resolve(import.meta.dirname, "..");
const parentCapability = Buffer.from(Array.from({ length: 32 }, (_, index) => index));
const clientNonce = Buffer.from(Array.from({ length: 32 }, (_, index) => 0x20 + index));
const helperNonce = Buffer.from(Array.from({ length: 32 }, (_, index) => 0x40 + index));
const replacementHelperNonce = Buffer.from(Array.from({ length: 32 }, (_, index) => 0x60 + index));
const helloBytes = Buffer.from(
  "{\"component\":\"discord_waifus\",\"nonce\":\"client\",\"protocol\":{\"major\":1,\"minor\":0}}",
  "utf8"
);
const helloAckBytes = Buffer.from(
  "{\"component\":\"ts_connect\",\"nonce\":\"helper\",\"protocol\":{\"major\":1,\"minor\":0}}",
  "utf8"
);
const replacementHelloAckBytes = Buffer.from(
  "{\"component\":\"ts_connect\",\"nonce\":\"replacement\",\"protocol\":{\"major\":1,\"minor\":0}}",
  "utf8"
);

function expectProtocolError(action: () => unknown, code: string): void {
  try {
    action();
  } catch (error) {
    expect(error).toBeInstanceOf(WipcProtocolError);
    expect(error).toMatchObject({ code });
    return;
  }
  throw new Error(`Expected WIPC protocol error ${code}.`);
}

function parentSession(): WipcParentAuthSession {
  return new WipcParentAuthSession({ parentCapability, clientNonce, helloBytes });
}

function helperSession(): WipcHelperAuthSession {
  return new WipcHelperAuthSession({ parentCapability });
}

describe("WIPC one-launch mutual authentication", () => {
  it("blocks traffic until both proofs succeed, then erases both capability copies", () => {
    const parent = parentSession();
    const helper = helperSession();
    expectProtocolError(() => parent.assertTrafficAllowed(), "frame_before_authentication");
    expectProtocolError(() => helper.assertTrafficAllowed(), "frame_before_authentication");

    const parentProof = parent.beginCandidate({ helperNonce, helloAckBytes });
    const helperProof = helper.authenticateCandidate({
      clientNonce,
      helperNonce,
      helloBytes,
      helloAckBytes,
      parentProof
    });
    parent.completeCandidate(helperProof);

    expect(parent.authenticated).toBe(true);
    expect(helper.authenticated).toBe(true);
    expect(parent.capabilityAvailable).toBe(false);
    expect(helper.capabilityAvailable).toBe(false);
    expect(() => parent.assertTrafficAllowed()).not.toThrow();
    expect(() => helper.assertTrafficAllowed()).not.toThrow();
  });

  it("rejects wrong and reflected helper proofs without consuming the launch capability", () => {
    for (const mode of ["wrong", "reflected"] as const) {
      const parent = parentSession();
      const parentProof = parent.beginCandidate({ helperNonce, helloAckBytes });
      const candidate = mode === "reflected" ? parentProof : Buffer.alloc(32, 0xff);
      expectProtocolError(() => parent.completeCandidate(candidate), "invalid_helper_proof");
      expect(parent.authenticated).toBe(false);
      expect(parent.capabilityAvailable).toBe(true);

      const helper = helperSession();
      const replacementParentProof = parent.beginCandidate({
        helperNonce: replacementHelperNonce,
        helloAckBytes: replacementHelloAckBytes
      });
      const replacementHelperProof = helper.authenticateCandidate({
        clientNonce,
        helperNonce: replacementHelperNonce,
        helloBytes,
        helloAckBytes: replacementHelloAckBytes,
        parentProof: replacementParentProof
      });
      parent.completeCandidate(replacementHelperProof);
      expect(parent.authenticated).toBe(true);
    }
  });

  it("rejects a replayed parent proof on a changed HELLO transcript", () => {
    const oldParentProof = deriveWipcParentProof({
      parentCapability,
      clientNonce,
      helperNonce,
      helloBytes,
      helloAckBytes
    });
    const replayClientNonce = Buffer.from(clientNonce);
    replayClientNonce[0] ^= 1;
    const helper = helperSession();
    expectProtocolError(() => helper.authenticateCandidate({
      clientNonce: replayClientNonce,
      helperNonce,
      helloBytes: Buffer.concat([helloBytes, Buffer.from(" ")]),
      helloAckBytes,
      parentProof: oldParentProof
    }), "invalid_parent_proof");
    expect(helper.authenticated).toBe(false);
    expect(helper.capabilityAvailable).toBe(true);
  });

  it("recovers after a socket-race impersonator and refuses every second client after success", () => {
    const parent = parentSession();
    const fakeHelperNonce = Buffer.alloc(32, 0x7f);
    const fakeHelloAck = Buffer.from(
      "{\"component\":\"ts_connect\",\"nonce\":\"socket-race\",\"protocol\":{\"major\":1,\"minor\":0}}"
    );
    const exposedParentProof = parent.beginCandidate({
      helperNonce: fakeHelperNonce,
      helloAckBytes: fakeHelloAck
    });
    expectProtocolError(
      () => parent.completeCandidate(exposedParentProof),
      "invalid_helper_proof"
    );

    const helper = helperSession();
    const parentProof = parent.beginCandidate({ helperNonce, helloAckBytes });
    const helperProof = helper.authenticateCandidate({
      clientNonce,
      helperNonce,
      helloBytes,
      helloAckBytes,
      parentProof
    });
    parent.completeCandidate(helperProof);

    expectProtocolError(
      () => parent.beginCandidate({ helperNonce, helloAckBytes }),
      "auth_capability_unavailable"
    );
    expectProtocolError(() => helper.authenticateCandidate({
      clientNonce,
      helperNonce,
      helloBytes,
      helloAckBytes,
      parentProof
    }), "auth_capability_unavailable");
  });

  it("fails closed on out-of-order operations and exact-width violations", () => {
    const parent = parentSession();
    expectProtocolError(() => parent.completeCandidate(Buffer.alloc(32)), "auth_sequence_error");
    parent.beginCandidate({ helperNonce, helloAckBytes });
    expectProtocolError(
      () => parent.beginCandidate({ helperNonce, helloAckBytes }),
      "auth_sequence_error"
    );
    expectProtocolError(() => parent.completeCandidate(Buffer.alloc(31)), "invalid_auth_width");
    expect(parent.capabilityAvailable).toBe(true);
  });
});

describe("WIPC V1 public authentication-session fixture", () => {
  it("is canonical and byte-stable", async () => {
    const actual = await readFile(
      path.join(
        repositoryRoot,
        "contracts",
        "remote",
        "v1",
        "fixtures",
        "crypto",
        "wipc-auth-session-v1.json"
      ),
      "utf8"
    );
    expect(actual).toBe(serializeCanonicalContractJson(createWipcAuthSessionV1Fixture()));
  });

  it("drives the TypeScript session accept/reject behavior", async () => {
    const actual = await readFile(
      path.join(
        repositoryRoot,
        "contracts",
        "remote",
        "v1",
        "fixtures",
        "crypto",
        "wipc-auth-session-v1.json"
      ),
      "utf8"
    );
    const fixture = JSON.parse(actual) as {
      parentCapabilityB64: string;
      parent: { clientNonceB64: string; helloBytesB64: string };
      candidates: Record<string, {
        clientNonceB64?: string;
        helloBytesB64?: string;
        helperNonceB64: string;
        helloAckBytesB64: string;
        parentProofB64: string;
        helperProofB64?: string;
        expectedParentError?: string;
        expectedHelperError?: string;
        expectedCapabilityState: "retained" | "erased";
      }>;
      rules: {
        trafficBeforeAuthenticationError: string;
        secondClientError: string;
        recoveryOrder: [string, string];
      };
    };
    const decode = (value: string) => Buffer.from(value, "base64url");
    const capability = decode(fixture.parentCapabilityB64);
    const baseClientNonce = decode(fixture.parent.clientNonceB64);
    const baseHello = decode(fixture.parent.helloBytesB64);
    const candidate = (name: string) => fixture.candidates[name];

    for (const name of ["wrongHelperProof", "reflectedParentProof"] as const) {
      const vector = candidate(name);
      const parent = new WipcParentAuthSession({
        parentCapability: capability,
        clientNonce: baseClientNonce,
        helloBytes: baseHello
      });
      expect(parent.beginCandidate({
        helperNonce: decode(vector.helperNonceB64),
        helloAckBytes: decode(vector.helloAckBytesB64)
      }).toString("base64url")).toBe(vector.parentProofB64);
      expectProtocolError(
        () => parent.completeCandidate(decode(vector.helperProofB64 ?? "")),
        vector.expectedParentError ?? ""
      );
      expect(parent.capabilityAvailable).toBe(vector.expectedCapabilityState === "retained");
    }

    const replay = candidate("replayedParentProof");
    const replayHelper = new WipcHelperAuthSession({ parentCapability: capability });
    expectProtocolError(() => replayHelper.authenticateCandidate({
      clientNonce: decode(replay.clientNonceB64 ?? ""),
      helperNonce: decode(replay.helperNonceB64),
      helloBytes: decode(replay.helloBytesB64 ?? ""),
      helloAckBytes: decode(replay.helloAckBytesB64),
      parentProof: decode(replay.parentProofB64)
    }), replay.expectedHelperError ?? "");
    expect(replayHelper.capabilityAvailable).toBe(true);

    const parent = new WipcParentAuthSession({
      parentCapability: capability,
      clientNonce: baseClientNonce,
      helloBytes: baseHello
    });
    expectProtocolError(
      () => parent.assertTrafficAllowed(),
      fixture.rules.trafficBeforeAuthenticationError
    );
    const [raceName, recoveryName] = fixture.rules.recoveryOrder;
    const race = candidate(raceName);
    parent.beginCandidate({
      helperNonce: decode(race.helperNonceB64),
      helloAckBytes: decode(race.helloAckBytesB64)
    });
    expectProtocolError(
      () => parent.completeCandidate(decode(race.helperProofB64 ?? "")),
      race.expectedParentError ?? ""
    );

    const recovery = candidate(recoveryName);
    const helper = new WipcHelperAuthSession({ parentCapability: capability });
    const recoveryParentProof = parent.beginCandidate({
      helperNonce: decode(recovery.helperNonceB64),
      helloAckBytes: decode(recovery.helloAckBytesB64)
    });
    expect(recoveryParentProof.toString("base64url")).toBe(recovery.parentProofB64);
    const recoveryHelperProof = helper.authenticateCandidate({
      clientNonce: baseClientNonce,
      helperNonce: decode(recovery.helperNonceB64),
      helloBytes: baseHello,
      helloAckBytes: decode(recovery.helloAckBytesB64),
      parentProof: recoveryParentProof
    });
    expect(recoveryHelperProof.toString("base64url")).toBe(recovery.helperProofB64);
    parent.completeCandidate(recoveryHelperProof);
    expect(parent.capabilityAvailable).toBe(false);
    expect(helper.capabilityAvailable).toBe(false);
    expectProtocolError(
      () => parent.beginCandidate({ helperNonce, helloAckBytes }),
      fixture.rules.secondClientError
    );
  });
});

File: contracts/remote/v1/conformance-go/conformance/wipc_auth_session_test.go
package conformance_test

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"

	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/vectors"
	"github.com/waifucave/discord-waifus/contracts/remote/v1/conformance-go/internal/wipc"
)

func authSessionFixtureBytes(t *testing.T) []byte {
	t.Helper()
	value, err := os.ReadFile(filepath.Join("..", "..", "fixtures", "crypto", "wipc-auth-session-v1.json"))
	if err != nil {
		t.Fatalf("read committed WIPC auth-session fixture: %v", err)
	}
	return value
}

func TestGoGeneratorMatchesCommittedWIPCAuthSessionV1Fixture(t *testing.T) {
	expected, err := vectors.BuildWIPCAuthSessionV1JSON()
	if err != nil {
		t.Fatalf("build WIPC auth-session fixture: %v", err)
	}
	if !bytes.Equal(authSessionFixtureBytes(t), expected) {
		t.Fatalf("Go-generated WIPC auth-session fixture differs from committed bytes")
	}
}

func newParentSessionFromFixture(
	t *testing.T,
	fixture vectors.WIPCAuthSessionV1Fixture,
) *wipc.ParentAuthSession {
	t.Helper()
	session, err := wipc.NewParentAuthSession(
		decodeBase64URL(t, fixture.ParentCapabilityB64),
		decodeBase64URL(t, fixture.Parent.ClientNonceB64),
		decodeBase64URL(t, fixture.Parent.HelloBytesB64),
	)
	if err != nil {
		t.Fatalf("new parent auth session: %v", err)
	}
	return session
}

func newHelperSessionFromFixture(
	t *testing.T,
	fixture vectors.WIPCAuthSessionV1Fixture,
) *wipc.HelperAuthSession {
	t.Helper()
	session, err := wipc.NewHelperAuthSession(decodeBase64URL(t, fixture.ParentCapabilityB64))
	if err != nil {
		t.Fatalf("new helper auth session: %v", err)
	}
	return session
}

func TestWIPCAuthSessionFixture(t *testing.T) {
	fixture, err := vectors.DecodeWIPCAuthSessionV1Fixture(authSessionFixtureBytes(t))
	if err != nil {
		t.Fatalf("decode auth-session fixture: %v", err)
	}

	valid, ok := fixture.Candidate("valid")
	if !ok {
		t.Fatalf("valid candidate is missing")
	}
	parent := newParentSessionFromFixture(t, fixture)
	helper := newHelperSessionFromFixture(t, fixture)
	if err := parent.AssertTrafficAllowed(); protocolErrorCode(t, err) != fixture.Rules.TrafficBeforeAuthenticationError {
		t.Fatalf("unexpected pre-auth parent error: %v", err)
	}
	parentProof, err := parent.BeginCandidate(
		decodeBase64URL(t, valid.HelperNonceB64),
		decodeBase64URL(t, valid.HelloAckBytesB64),
	)
	if err != nil {
		t.Fatalf("begin valid candidate: %v", err)
	}
	if !bytes.Equal(parentProof, decodeBase64URL(t, valid.ParentProofB64)) {
		t.Fatalf("valid parent proof differs from fixture")
	}
	helperProof, err := helper.AuthenticateCandidate(wipc.HelperCandidate{
		ClientNonce: decodeBase64URL(t, fixture.Parent.ClientNonceB64),
		HelperNonce: decodeBase64URL(t, valid.HelperNonceB64),
		Hello:       decodeBase64URL(t, fixture.Parent.HelloBytesB64),
		HelloAck:    decodeBase64URL(t, valid.HelloAckBytesB64),
		ParentProof: parentProof,
	})
	if err != nil {
		t.Fatalf("authenticate valid helper: %v", err)
	}
	if !bytes.Equal(helperProof, decodeBase64URL(t, valid.HelperProofB64)) {
		t.Fatalf("valid helper proof differs from fixture")
	}
	if err := parent.CompleteCandidate(helperProof); err != nil {
		t.Fatalf("complete valid parent: %v", err)
	}
	if !parent.Authenticated() || !helper.Authenticated() {
		t.Fatalf("valid sessions did not authenticate")
	}
	if parent.CapabilityAvailable() || helper.CapabilityAvailable() {
		t.Fatalf("successful sessions retained a capability")
	}
	if err := parent.AssertTrafficAllowed(); err != nil {
		t.Fatalf("authenticated parent blocked traffic: %v", err)
	}

	for _, name := range []string{"wrongHelperProof", "reflectedParentProof"} {
		candidate, ok := fixture.Candidate(name)
		if !ok {
			t.Fatalf("candidate %q is missing", name)
		}
		parent := newParentSessionFromFixture(t, fixture)
		if _, err := parent.BeginCandidate(
			decodeBase64URL(t, candidate.HelperNonceB64),
			decodeBase64URL(t, candidate.HelloAckBytesB64),
		); err != nil {
			t.Fatalf("begin %s: %v", name, err)
		}
		err := parent.CompleteCandidate(decodeBase64URL(t, candidate.HelperProofB64))
		if got := protocolErrorCode(t, err); got != candidate.ExpectedParentError {
			t.Fatalf("%s error = %q, want %q", name, got, candidate.ExpectedParentError)
		}
		if !parent.CapabilityAvailable() || parent.Authenticated() {
			t.Fatalf("%s consumed capability or authenticated", name)
		}
	}

	replay, ok := fixture.Candidate("replayedParentProof")
	if !ok {
		t.Fatalf("replayed parent candidate is missing")
	}
	replayHelper := newHelperSessionFromFixture(t, fixture)
	_, err = replayHelper.AuthenticateCandidate(wipc.HelperCandidate{
		ClientNonce: decodeBase64URL(t, replay.ClientNonceB64),
		HelperNonce: decodeBase64URL(t, replay.HelperNonceB64),
		Hello:       decodeBase64URL(t, replay.HelloBytesB64),
		HelloAck:    decodeBase64URL(t, replay.HelloAckBytesB64),
		ParentProof: decodeBase64URL(t, replay.ParentProofB64),
	})
	if got := protocolErrorCode(t, err); got != replay.ExpectedHelperError {
		t.Fatalf("replay error = %q, want %q", got, replay.ExpectedHelperError)
	}
	if !replayHelper.CapabilityAvailable() {
		t.Fatalf("replayed parent proof consumed helper capability")
	}
}

func TestWIPCAuthSessionSocketRaceRecoveryAndSecondClient(t *testing.T) {
	fixture, err := vectors.DecodeWIPCAuthSessionV1Fixture(authSessionFixtureBytes(t))
	if err != nil {
		t.Fatalf("decode auth-session fixture: %v", err)
	}
	race, ok := fixture.Candidate(fixture.Rules.RecoveryOrder[0])
	if !ok {
		t.Fatalf("race candidate is missing")
	}
	recovery, ok := fixture.Candidate(fixture.Rules.RecoveryOrder[1])
	if !ok {
		t.Fatalf("recovery candidate is missing")
	}
	parent := newParentSessionFromFixture(t, fixture)
	if _, err := parent.BeginCandidate(
		decodeBase64URL(t, race.HelperNonceB64),
		decodeBase64URL(t, race.HelloAckBytesB64),
	); err != nil {
		t.Fatalf("begin race candidate: %v", err)
	}
	err = parent.CompleteCandidate(decodeBase64URL(t, race.HelperProofB64))
	if got := protocolErrorCode(t, err); got != race.ExpectedParentError {
		t.Fatalf("race error = %q, want %q", got, race.ExpectedParentError)
	}

	helper := newHelperSessionFromFixture(t, fixture)
	parentProof, err := parent.BeginCandidate(
		decodeBase64URL(t, recovery.HelperNonceB64),
		decodeBase64URL(t, recovery.HelloAckBytesB64),
	)
	if err != nil {
		t.Fatalf("begin recovery: %v", err)
	}
	helperProof, err := helper.AuthenticateCandidate(wipc.HelperCandidate{
		ClientNonce: decodeBase64URL(t, fixture.Parent.ClientNonceB64),
		HelperNonce: decodeBase64URL(t, recovery.HelperNonceB64),
		Hello:       decodeBase64URL(t, fixture.Parent.HelloBytesB64),
		HelloAck:    decodeBase64URL(t, recovery.HelloAckBytesB64),
		ParentProof: parentProof,
	})
	if err != nil {
		t.Fatalf("helper recovery: %v", err)
	}
	if err := parent.CompleteCandidate(helperProof); err != nil {
		t.Fatalf("parent recovery: %v", err)
	}

	_, err = parent.BeginCandidate(
		decodeBase64URL(t, recovery.HelperNonceB64),
		decodeBase64URL(t, recovery.HelloAckBytesB64),
	)
	if got := protocolErrorCode(t, err); got != fixture.Rules.SecondClientError {
		t.Fatalf("parent second-client error = %q, want %q", got, fixture.Rules.SecondClientError)
	}
	_, err = helper.AuthenticateCandidate(wipc.HelperCandidate{
		ClientNonce: decodeBase64URL(t, fixture.Parent.ClientNonceB64),
		HelperNonce: decodeBase64URL(t, recovery.HelperNonceB64),
		Hello:       decodeBase64URL(t, fixture.Parent.HelloBytesB64),
		HelloAck:    decodeBase64URL(t, recovery.HelloAckBytesB64),
		ParentProof: parentProof,
	})
	if got := protocolErrorCode(t, err); got != fixture.Rules.SecondClientError {
		t.Fatalf("helper second-client error = %q, want %q", got, fixture.Rules.SecondClientError)
	}
}

func TestWIPCAuthSessionSequenceErrors(t *testing.T) {
	fixture, err := vectors.DecodeWIPCAuthSessionV1Fixture(authSessionFixtureBytes(t))
	if err != nil {
		t.Fatalf("decode auth-session fixture: %v", err)
	}
	valid, ok := fixture.Candidate("valid")
	if !ok {
		t.Fatalf("valid candidate is missing")
	}
	parent := newParentSessionFromFixture(t, fixture)
	err = parent.CompleteCandidate(make([]byte, 32))
	if got := protocolErrorCode(t, err); got != fixture.Rules.CompletionWithoutCandidateError {
		t.Fatalf("completion error = %q, want %q", got, fixture.Rules.CompletionWithoutCandidateError)
	}
	if _, err := parent.BeginCandidate(
		decodeBase64URL(t, valid.HelperNonceB64),
		decodeBase64URL(t, valid.HelloAckBytesB64),
	); err != nil {
		t.Fatalf("begin first candidate: %v", err)
	}
	_, err = parent.BeginCandidate(
		decodeBase64URL(t, valid.HelperNonceB64),
		decodeBase64URL(t, valid.HelloAckBytesB64),
	)
	if got := protocolErrorCode(t, err); got != fixture.Rules.CandidateAlreadyActiveError {
		t.Fatalf("duplicate candidate error = %q, want %q", got, fixture.Rules.CandidateAlreadyActiveError)
	}
	if !parent.CapabilityAvailable() {
		t.Fatalf("sequence error consumed parent capability")
	}

	helper := newHelperSessionFromFixture(t, fixture)
	err = helper.AssertTrafficAllowed()
	if got := protocolErrorCode(t, err); got != fixture.Rules.TrafficBeforeAuthenticationError {
		t.Fatalf("helper traffic error = %q, want %q", got, fixture.Rules.TrafficBeforeAuthenticationError)
	}
}
